core/domain/exp_domain_test.py

```diff
@@ -717,14 +717,18 @@ def test_validation(self):
         init_state.update_interaction_default_outcome(default_outcome.to_dict())
         exploration.validate()
 
-        init_state.update_interaction_solution({
+        solution_dict = {
             'answer_is_exclusive': True,
             'correct_answer': 'hello_world!',
             'explanation': {
                 'content_id': 'solution',
                 'html': 'hello_world is a string'
-                }
-        })
+            }
+        }
+        solution = state_domain.Solution.from_dict(
```

Comment:
Perhaps here you could construct the Solution object directly, using the constructor. Ditto elsewhere. In general we should try to avoid using dicts except at endpoints of the app.

Reply:
Done.
The service falls under the "endpoint case" (naming sort of states that) right?

Reply:
No, I don't think so -- "endpoint" here means the controller layer, i.e. the part in core.controllers that is just before self.render_json(), since that is the point at which we need to serialize stuff before sending it to the frontend. (Sorry for not being clear before.)